src/python_routing_v02/diffusive_utils.py

```diff
@@ -0,0 +1,654 @@
+            , qlat_data            
+            , qlat_g
+            ):   
+    
+    frj= -1
+    for x in range(mx_jorder_tw,-1,-1):
+        for head_segment, reach in ordered_reaches[x]:                  
+            seg_list= reach['segments_list']
+            ncomp= reach['number_segments']          
+            frj= frj+1     
+            for seg in range(0,ncomp):                               
+                segID= seg_list[seg]                
+                for tsi in range (0,nts_ql_g):
+                    if seg<ncomp-1:        
+                        
+                        idx_segID = np.where(geo_index == segID)
+                        idx_par = np.where(geo_cols == "dx")
+                        
+                        tlf= qlat_data[idx_segID,tsi]  # [m^3/sec]
+                        dx=  geo_data[idx_segID,idx_par] # [meter]
+                        qlat_g[tsi,seg,frj]= tlf/dx   #[m^2/sec]
+                        
+                    else:
+                        qlat_g[tsi,seg,frj]= 0.0 # seg=ncomp is actually for bottom node in Fotran code.
+                                                     #And, lateral flow enters between adjacent nodes.
+       
+    return 
+
+#------------------------------------------------------------------------------------------
+# upstream boundary condition mapping between Python and Fortran
+# 
+# Assumption: top segments of headbasin reaches get their own lateral flows as 
+#             upstream boundary conditions while making these qlateral zeroes
+#------------------------------------------------------------------------------------------
+def fp_ubcd_map(frnw_g
+            , pynw
+            , nts_ub_g
+            , nrch_g
+            , geo_index
+            , qlat_data
+            , qlat_g
+            ):   
+    
+    ubcd_g=np.zeros((nts_ub_g, nrch_g)) 
+    frj= -1
+    for frj in range(nrch_g):
+        if frnw_g[frj,2]==0: # the number of upstream reaches is zero.
+            head_segment=pynw[frj]
+            for tsi in range(0,nts_ub_g):
+                
+                idx_segID = np.where(geo_index == head_segment)
+                
+                ubcd_g[tsi, frj]= qlat_data[idx_segID,tsi] # [m^3/s]
+                qlat_g[tsi,0,frj]=0.0                   
+    
+    return ubcd_g 
+                        
+#------------------------------------------------------------------------------------------
+#       downstream boundary condition mapping between Python and Fortran
+# 
+#   ** In this mapping, USGS observed stage data at TW is used.
+#------------------------------------------------------------------------------------------
+def fp_dbcd_map(usgsID2tw
+                , usgssDT
+                , usgseDT
+                , usgspCd
+                ):
+   
+    
+    # ** 1) downstream stage (here, lake elevation) boundary condition
+    #from nwis_client.iv import IVDataService
+    from evaluation_tools.nwis_client.iv import IVDataService
```

Comment:
Should update to the `hydrotools` package.  Really just need to change this package here and make sure it is installed, so probably should add `hydrotools` to requirements.txt and/or setup.py

Reply:
@jameshalgren this might be the what you were referring to in your comment on USGS lib.  You can find it [here](https://pypi.org/project/hydrotools.nwis-client/) on pypi or [here](https://github.com/NOAA-OWP/hydrotools/tree/main/python/nwis_client) on github.